### PR TITLE
Fix/java api

### PR DESCRIPTION
**Pull request by johanfylling**
* Correcting the order of `input` and `data` arguments to `Plan.eval()`
* Sanitizing collection types in `input` and `data`

### CODE REVIEW

File: api/src/test/java/by/borge/jarl/JarlTests.java [hunk 1]
```diff
@@ -1,25 +1,58 @@
...
+                Arguments.of(null, "bar"),
+                Arguments.of("foo", "bar"),
+                Arguments.of(1337, 42),
+                Arguments.of(13.37, 4.2),
+                Arguments.of(true, false),
+                Arguments.of(false, true),
+                Arguments.of(List.of(), List.of()),
+//                Arguments.of(
+//                        List.of(1, "two", 3.0),
+//                        List.of(4, "five", 6.0)),
```
Comment (johanfylling) [inline api/src/test/java/by/borge/jarl/JarlTests.java:29]: This is interesting: somehow, lists are reversed

```
expected: <[1, two, 3.0]> but was: <[3.0 "two" 1]>
Expected :[1, two, 3.0]
Actual   :[3.0 "two" 1]
```

File: core/src/main/clj/jarl/api.clj [hunk 2]
```diff
@@ -15,11 +16,18 @@
 (defn -plan-init [info plan entry-point]
   [[info] [info plan entry-point]])
 
-(defn -plan-eval [^by.borge.jarl.internal.InternalPlanImpl this input data]
+(defn- sanitize [data]
+  (cond
+    (instance? Map data) (reduce-kv (fn [d k v] (assoc d k (sanitize v))) {} data)
+    (instance? List data) (reduce (fn [d v] (conj d (sanitize v))) '() data)
```
[the remark this answers is not in the document]
Comment (johanfylling) [reply core/src/main/clj/jarl/api.clj]: Transforming Java Lists into clj vectors instead.

File: api/src/test/java/by/borge/jarl/JarlTests.java [hunk 3]
```diff
@@ -1,25 +1,58 @@
...
+                                "map", Map.of("a", "b")),
+                        Map.of(
+                                "bar", "foo",
+                                "two", 2,
+                                "list", List.of(4, 5, 6),
+                                "map", Map.of("c", "d")))
+        );
+    }
+
+    @ParameterizedTest
```
Comment (anderseknert) [inline api/src/test/java/by/borge/jarl/JarlTests.java:44]: I'm not following here.. I'm taking it this is some JUnit magic :) The method below will be called once for each item in the stream, is that right?

File: api/src/main/java/by/borge/jarl/Plan.java [hunk 4]
```diff
@@ -17,11 +17,11 @@ public class Plan {
...
      * Evaluate this plan given <code>input</code> and <code>data</code>.
      *
-     * @param input the input value
      * @param data the data document
+     * @param input the input value
      * @return a {@link ResultSet} representing the evaluation result of this Plan
      */
-    public ResultSet eval(Object input, Map<String, ?> data) {
-        return ResultSet.fromResultsList(plan.eval(input, data));
+    public ResultSet eval(Map<String, ?> data, Object input) {
```
Comment (anderseknert) [inline api/src/main/java/by/borge/jarl/Plan.java:24]: I actually prefer the previous order, as more eval calls depend on input than data... if we use input and data in that order, we could have overloads like:

```java
public ResultSet eval()
public ResultSet eval(Object input)
public ResultSet eval(Object input, Map<String, ?> data)
```
This order is also what you used in the test data, so I think that's what comes natural. It's also the order used in the plans, i.e. local 0 (input) and 1 (data)

File: api/src/test/java/by/borge/jarl/JarlTests.java (hunk 3 above)
Comment (johanfylling) [reply api/src/test/java/by/borge/jarl/JarlTests.java:44]: Yep. It's all tied together by the test method and the argument factory method having the same name.

I considered using spock/groovy instead, as I find that immensely more readable, but decided against it as this is mainly to test the Java API, and having an extra layer of language indirection in the tests might introduce issues, or even hide issues through the magic of groovy.

File: api/src/main/java/by/borge/jarl/Plan.java (hunk 4 above)
Comment (johanfylling) [reply api/src/main/java/by/borge/jarl/Plan.java:24]: True. But then we should probably trickle that through all the way, and change the order of the args to the plans generated by `jarl.parser.make-plan`, and make it all consistent.

Which'll likely involve changing a lot of tests 😅 , but I'm up for it.

Comment (anderseknert) [reply api/src/main/java/by/borge/jarl/Plan.java:24]: Oh, I hadn't thought about that.. but yes, if you're up for it, that'd be great! 👍

Comment (anderseknert) [reply api/src/main/java/by/borge/jarl/Plan.java:24]: Don't forget about docs too, in case we've documented the other order there

Comment (johanfylling) [reply api/src/main/java/by/borge/jarl/Plan.java:24]: Yea, the clj API will be part of the change 👍 .

**Review by anderseknert — APPROVED** [review]
Thanks!

**Pull request merged**